static/app/views/explore/toolbar/toolbarGroupBy.tsx

```diff
@@ -60,61 +74,203 @@ export function ToolbarGroupBy({disabled}: ToolbarGroupByProps) {
     ];
   }, [groupBys, tags]);
 
-  const addGroupBy = useCallback(() => {
-    setGroupBys([...groupBys, '']);
-  }, [setGroupBys, groupBys]);
-
-  const setGroupBy = useCallback(
-    (i: number, {value}: SelectOption<Field>) => {
-      const newGroupBys = groupBys.slice();
-      newGroupBys[i] = value;
-      setGroupBys(newGroupBys);
-    },
-    [setGroupBys, groupBys]
-  );
+  const {
+    editableColumns,
+    insertColumn,
+    updateColumnAtIndex,
+    deleteColumnAtIndex,
+    swapColumnsAtIndex,
+  } = useDragNDropColumns({columns: groupBys});
 
-  const deleteGroupBy = useCallback(
-    (index: number) => {
-      const newGroupBys = groupBys.filter((_, orgIndex) => index !== orgIndex);
-      setGroupBys?.(newGroupBys);
-    },
-    [setGroupBys, groupBys]
-  );
+  useEffect(() => {
+    setGroupBys(editableColumns.map(({column}) => column ?? ''));
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [editableColumns]);
```

Comment:
This API is very strange to require that the caller of `useDragNDropColumns` to have to explicitly propagate the changes to itself. I would think this is something handled by the hook for you.

So perhaps an API that looks more like

```typescript
const {
  insertColumn,
  updateColumnAtIndex,
  deleteColumnAtIndex,
  swapColumnsAtIndex,
} = useDragNDropColumns({columns: groupBys, setColumns: setGroupBys});
```

to be more ergonomic and allow the hook to update the state of the array as needed.